### PR TITLE
Support google.golang.org/grpc v1.69.0

### LINKED ISSUE #1421: Support `google.golang.org/grpc` `v1.69.0`
The function `google.golang.org/grpc/internal/transport.(*http2Server).WriteStatus` was renamed to `function google.golang.org/grpc/internal/transport.(*http2Server).writeStatus`: https://github.com/grpc/grpc-go/blob/317271b232677b7869576a49855b01b9f4775d67/internal/transport/http2_server.go#L1045

_Originally posted by @MrAlias in https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1416#issuecomment-2540145321_

### PR DESCRIPTION
**Pull request by MrAlias**
Resolve #1421
Resolve https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1432

In `google.golang.org/grpc` `v1.69.0` the `http2Server.writeStatus` and `Server.handleStream` method signatures changed:

- `http2Server.writeStatus`
  - `v1.68.1`: https://github.com/grpc/grpc-go/blob/d6a777f952c77822f0190dff71b1fe8fe250538c/internal/transport/http2_server.go#L1049
  - `v1.69.0`: https://github.com/grpc/grpc-go/blob/317271b232677b7869576a49855b01b9f4775d67/internal/transport/http2_server.go#L1045
- `Server.handleStream`
  - `v1.68.1`: https://github.com/grpc/grpc-go/blob/d6a777f952c77822f0190dff71b1fe8fe250538c/server.go#L1741
  - `v1.69.0`: https://github.com/grpc/grpc-go/blob/317271b232677b7869576a49855b01b9f4775d67/server.go#L1735

In both cases, they are now passed [`*transport.ServerStream`](https://github.com/grpc/grpc-go/blob/317271b232677b7869576a49855b01b9f4775d67/internal/transport/server_stream.go#L34) instead of [`*transport.Stream`](https://github.com/grpc/grpc-go/blob/d6a777f952c77822f0190dff71b1fe8fe250538c/internal/transport/transport.go#L289). The `v1.69.0` version of [`transport.Stream`](https://github.com/grpc/grpc-go/blob/317271b232677b7869576a49855b01b9f4775d67/internal/transport/transport.go#L287-L309) hasn't changed, just how it is referenced. In `v1.69.0` the [`*Stream` is embedded in `ServerStream`](https://github.com/grpc/grpc-go/blob/317271b232677b7869576a49855b01b9f4775d67/internal/transport/server_stream.go#L34-L36).

This PR handles this divergence and support both code paths.

## Split uprobes

To reduce complexity of the uprobes by adding another `volatile` flag variable, this splits functionality of the uprobes. A new `PackageConstraints` field is added to the `probe.Uprobe` to provide scope of when the probe should be loaded and the behavior when it is not loaded.

This "behavior when not loaded" concept overlaps with then already existing `Optional` field of the `Probe`. This is replace with a more customizable `FailureMode` type that describes the behavior.

The common functionality of the uprobes are abstracted to inlined functions.

## Event type updates

The event type now reports if the status was set.

## Go context parsing

The Go `context.Context` is held within the `Stream.ctx` field. This is not directly accessible via passed arguments of the instrumented methods; it is only accessable via `ServerStream.Stream.ctx`, there is an added level of indirection. This added level of indirection means for the updated code paths our header file functionality is not used.

## `google.golang.org/grpc` `v1.68.1`

This was manually tested against `google.golang.org/grpc` `v1.68.1` by downgrading the version of `grpc` used in our e2e tests and tested.

### CODE REVIEW

File: internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go [hunk 1]
```diff
@@ -110,6 +129,31 @@ func New(logger *slog.Logger, version string) probe.Probe {
 					Sym:         "google.golang.org/grpc.(*Server).handleStream",
 					EntryProbe:  "uprobe_server_handleStream",
 					ReturnProbe: "uprobe_server_handleStream_Returns",
+					PackageConstrainsts: []probe.PackageConstrainst{
+						{
+							Package: "google.golang.org/grpc",
+							Constraints: version.MustConstraints(
+								version.NewConstraint(fmt.Sprintf(
+									"> %s, < %s", writeStatusMin, writeStatusV2),
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go:137]: Right. It looks like I just copy pasted when I should not have. I'll updated. :+1:

File: internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf/probe.bpf.c [hunk 2]
```diff
@@ -61,47 +62,65 @@ volatile const u64 frame_fields_pos;
...
+//   - ctx: the pt_regs passed to the uprobe function
+//   - stream_ptr: pointer to the transport.Stream tracking the stream
+//   - go_context: the parsed Go context.Context
+//
+// Returns 0 on success, otherwise a negative error value in case of failure.
+static __always_inline int handleStream(struct pt_regs *ctx, void *stream_ptr, struct go_iface *go_context) {
+    if (ctx == NULL) {
+        bpf_printk("grpc:server:handleStream: NULL ctx");
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf/probe.bpf.c]: Updated. PTAL.

File: internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrAlias) [reply internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go]: Updated. PTAL.

File: internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf/probe.bpf.c [hunk 3]
```diff
@@ -112,39 +131,196 @@ int uprobe_server_handleStream(struct pt_regs *ctx)
...
+//   - ctx: the pt_regs passed to the uprobe function
+//   - status_ptr: pointer to the status.Stream holding the status info
+//   - go_context: the parsed Go context.Context
+//
+// Returns 0 on success, otherwise a negative error value in case of failure.
+static __always_inline int writeStatus(struct pt_regs *ctx, void *status_ptr, struct go_iface *go_context) {
+    if (ctx == NULL) {
+        bpf_printk("grpc:server:writeStatus: NULL ctx");
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf/probe.bpf.c]: Updated. PTAL.

File: internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf/probe.bpf.c [hunk 4]
```diff
@@ -112,39 +131,196 @@ int uprobe_server_handleStream(struct pt_regs *ctx)
...
+    req_ptr->has_status = true;
+
+    return 0;
+}
+
+// This instrumentation attaches uprobe to the following function:
+// func (s *Server) handleStream(t transport.ServerTransport, stream *transport.Stream, trInfo *traceInfo)
+//
+// This is only compatible with versions < 1.40 and > 1.69.0 of the Server.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf/probe.bpf.c]: Updated. PTAL.

**Review by RonFed — COMMENTED** [review]
@MrAlias Do you think we should do the test we talked about yesterday to make sure the context reading is OK? (creating a child span for the gRPC server span)

This looks great to me

**Comment by MrAlias** [comment]
> @MrAlias Do you think we should do the test we talked about yesterday to make sure the context reading is OK? (creating a child span for the gRPC server span)
> 
> This looks great to me

I was tracking that work in https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1442.

I was unsure if we wanted to add it here as this PR is already quite large. I can add it as a follow up, and even add it in a branch based on this branch to validate this will work. Or, I can just add it to this PR. What is your preference @RonFed?

**Comment by MrAlias** [comment]
@RonFed please take a look at https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1468. Let me know if you think that should be combined with this PR.

**Comment by RonFed** [comment]
I think we can have the test in the same PR to make sure the context reading is correct. But we can also split, whatever you prefer

**Review by RonFed — APPROVED** [review]
🥇

**Pull request merged**